drivers/rtc/rtc_nxp_irtc.c

```diff
@@ -0,0 +1,502 @@
+				    RTC_ALM_DAYS_ALM_DAY_SHIFT);
+		*mask |= RTC_ALARM_TIME_MASK_MONTHDAY;
+	}
+
+	if ((curr_alarm_mask & RTC_ALARM_TIME_MASK_MONTH) == RTC_ALARM_TIME_MASK_MONTH) {
+		timeptr->tm_mon = ((irtc_reg->ALM_YEARMON & RTC_ALM_YEARMON_ALM_MON_MASK) >>
+				   RTC_ALM_YEARMON_ALM_MON_SHIFT) -
+				  1;
+		*mask |= RTC_ALARM_TIME_MASK_MONTH;
+	}
+
+	if ((curr_alarm_mask & RTC_ALARM_TIME_MASK_YEAR) == RTC_ALARM_TIME_MASK_YEAR) {
+		timeptr->tm_year =
+			(int8_t)((irtc_reg->ALM_YEARMON & RTC_ALM_YEARMON_ALM_YEAR_MASK) >>
+				 RTC_ALM_YEARMON_ALM_YEAR_SHIFT) +
+			RTC_NXP_IRTC_YEAR_OFFSET;
+		*mask |= RTC_ALARM_TIME_MASK_YEAR;
+	}
+
+	return 0;
+}
+
+static int nxp_irtc_alarm_is_pending(const struct device *dev, uint16_t id)
+{
+	struct nxp_irtc_data *data = dev->data;
+	int ret = data->is_alarm_pending;
+
+	if (id != 0) {
+		return -EINVAL;
+	}
+
+	data->is_alarm_pending = 0;
+
+	return ret;
+}
+
+static int nxp_irtc_alarm_set_callback(const struct device *dev, uint16_t id,
+				       rtc_alarm_callback callback, void *user_data)
+{
+	struct nxp_irtc_data *data = dev->data;
+
+	if (id != 0) {
+		return -EINVAL;
+	}
+
+	data->alarm_callback = callback;
+	data->alarm_user_data = user_data;
+
+	return 0;
+}
+
+#endif /* CONFIG_RTC_ALARM */
+#if defined(CONFIG_RTC_UPDATE) || defined(__DOXYGEN__)
+
+static int nxp_irtc_update_set_callback(const struct device *dev, rtc_update_callback callback,
+					void *user_data)
+{
+	struct nxp_irtc_data *data = dev->data;
+
+	data->update_callback = callback;
+	data->update_user_data = user_data;
+
+	return 0;
+}
+
+#endif /* CONFIG_RTC_UPDATE */
+#if defined(CONFIG_RTC_CALIBRATION) || defined(__DOXYGEN__)
+
+static int nxp_irtc_set_calibration(const struct device *dev, int32_t calibration)
+{
+	struct nxp_irtc_data *data = dev->data;
+
+	data->calibration = calibration;
+
+	return 0;
+}
+
+static int nxp_irtc_get_calibration(const struct device *dev, int32_t *calibration)
+{
+	struct nxp_irtc_data *data = dev->data;
+
+	*calibration = data->calibration;
+
+	return 0;
+}
+
+#endif /* CONFIG_RTC_CALIBRATION */
+
+static int nxp_irtc_init(const struct device *dev)
+{
+	const struct nxp_irtc_config *config = dev->config;
+	RTC_Type *irtc_reg = config->base;
+
+	nxp_irtc_unlock_registers(irtc_reg);
+
+	/* reset all counter values */
+	irtc_reg->CTRL = RTC_CTRL_SWR_MASK;
+
+	/* set the control register bits */
+	uint32_t bits_to_set = RTC_CTRL_CLK_SEL(config->clock_select) |
+			       RTC_CTRL_CLKO_DIS(!config->is_output_clock_enabled);
+	irtc_reg->CTRL = bits_to_set;
+
+	config->irq_config_func(dev);
+
+	return 0;
+}
+
+static int nxp_irtc_isr(const struct device *dev)
+{
+#ifdef CONFIG_RTC_ALARM
+	const struct nxp_irtc_config *config = dev->config;
+	RTC_Type *irtc_reg = config->base;
+	struct nxp_irtc_data *data = dev->data;
+	uint32_t key = irq_lock();
+
+	nxp_irtc_unlock_registers(irtc_reg);
+	irtc_reg->ISR = irtc_reg->ISR;
+
+	if (data->alarm_callback) {
```

Comment:
The ISR is shared between alarm and update, we must check if the alarm or update caused the ISR before calling callbacks or setting alarm_pending flag.

Reply:
I did not mean to enable the Update section, I don't believe this peripheral has an Update functionality, sorry about the confusion.